### ISSUE TITLE
General FAQ

### ISSUE THREAD

**Issue by yyx990803** [vuejs/vue#96]
Use this thread for simple, quick questions to avoid cluttering the issue list. Alternatively you can try the #vuejs IRC channel on freenode.

Also, read the wiki [FAQ](https://github.com/yyx990803/vue/wiki/FAQ) first.

**Comment by larafale** [comment]
hello @yyx990803 , first of all, thx for this beauty, been using vue for all my project now !
one little question, i have an object where i want to set a dynamic key like

```$vue.$data.stats[myDynamicKey] = "somedata"```

this does not trigger view update. I know that on arrays you can use $set to explicitly call setters, but what about objects? can't find anything on your docs.

any clues ?
thx

**Comment by yyx990803** [comment]
@larafale it's the same for objects: `vue.$data.stats.$set(myDynamicKey, "someData")`

**Comment by larafale** [comment]
that's weird, 

```$vue.$data.stats[myDynamicKey] = "somedata"``` 
works if i predefine the keys in the vue data object, like
```
$vue = new Vue({
  el: '#content',
  data: {
    stats: {
      myDynamicKey1: {},
      myDynamicKey2: {},
    }
  }
})
```

but if i go like this
```
$vue = new Vue({
  el: '#content',
  data: {
    stats: {}
  }
})
```
nothing triggers, and concerning 
```$vue.$data.stats.$set(myDynamicKey, "someData")``` 
it always return undefined for the $set method with or without the object keys defined in data.stats

i've used $set on arrays before, worked good, but now its on object, and having the problem. i'm using v0.11.5

**Comment by yyx990803** [comment]
The guide talks about it a little bit here: http://vuejs.org/guide/#Model - basically Vue cannot detect new property being added unless using `$set`.
`$set` doesn't return anything, so `undefined` is expected, but it should trigger view changes.

**Comment by larafale** [comment]
yep i get it :) but here it's not the function that returns undefined
it's the $set function that is undefined
so when i call 
```$vue.$data.stats.$set(myDynamicKey, "someData")``` 
i get  Uncaught TypeError: undefined is not a function.  

if i define stats as an array, $set is not undefined, but here i need stats to be an object.


thx evan

**Comment by yyx990803** [comment]
Ah sorry, my bad, it's `$add`. After the initial `$add` call you can just set it like normal properties.

**Comment by larafale** [comment]
ok evan, thanks you rock solid :) 
so much impressed by your work, so clean and well designed, i've been deep into other frameworks like angular etc for years, but vue is masterclass. only react can compete, but it's for larger scale apps.
keep on !!

**Comment by dutgriff** [comment]
I think I understand this right. Updating $data is run asynchronously. For instance calling `this.items = [{...}];` updates the vm.$data.items then does a few things after that all asynchronously. So calling jQuery to check out an element right after changing the data might not return what one would expect because the DOM hasn't been updated just yet. Take [this fiddle](http://jsfiddle.net/mn54bq4f/) for example. 

*The question:* How can I wait until the items propagate entirely before running the jQuery?

The real-world case I am faced with is loading select options then calling select2 on them. When I call Select2 it fits the width of the select element prior to the $data.items loading so the select box stays really small. ~~Another damn Select2 question.~~ 

You can see in [this fiddle](http://jsfiddle.net/mn54bq4f/1) running the jQuery in a `Vue.nextTick` will wait long enough in the fiddle. This isn't the case for my app with Select2 in it. I guess it is because the app has a lot more processing going on and the DOM isn't updated after just 1 "tick".

**Comment by yyx990803** [comment]
@dutgriff see http://vuejs.org/api/global-api.html#Vue-nextTick\(_callback_\)

Call `Vue.nextTick(cb)` right after you changed the data and DOM will have been updated in the callback.

**Comment by dutgriff** [comment]
@yyx990803 Darn! You are right. As mentioned in my previous comment I tried that and it didn't work. There must be something going on in my app that I am missing. [This fiddle](http://jsfiddle.net/mn54bq4f/2/) shows nextTick working with a select2 example. Hmmm.
######Side note: Thank you so much for your dedication for Vue.js. You are really killing it man!

**Comment by jbruni** [comment]
I've deleted a key from an object used at `v-repeat`, but the DOM does not reflect the change.

HTML:

    <li v-repeat="tenants" class="list-group-item">things here</li>

JS:

    // Vue data snippet:
    data: {
        tenants: {
            '12': [{id: 5, name: 'John'}, {id: 7, name: 'Joe'}],
            '15': [{id: 25, name: 'Sarah'}, {id: 27, name: 'Mary'}]
        }
    }

    // Somewhere in code (inside a vm method)
    delete vm.tenants['12'];

The `<li>` element for the `tenants['12']` is kept unchanged, although the object is modified.

**Comment by yyx990803** [comment]
@jbruni you need to use `vm.$delete(key)`. The reason is mentioned in the docs: http://vuejs.org/guide/list.html#Iterating_Through_An_Object (scroll down for the red tip section)

**Comment by jbruni** [comment]
@yyx990803 - Perfect! Thanks a lot. :+1: Including for the amazing ultra-fast-speed support.

**Comment by kikyous** [comment]
Accroding
> It is also possible to expose $data as a prop. The passed in value must be an Object and will replace the component’s default $data.

```  
<component is="{{currentView}}" $data={{_data}}></component>
```
but this raise a error

**Comment by yyx990803** [comment]
@kikyous if you believe it's a bug, please open a separate issue with a reproduction.

**Comment by yyx990803** [comment]
@kikyous btw it seems you are trying to bind to unobserved data (prefixed with `_`), that won't work.

**Comment by abhijitghogre** [comment]
It is recommended in the best practices section that al the properties of model objects should be defined!

In my case I have a huge object with string and array values. I set all these fields using direct assignment in ajax success.

When I display them in my view, I get console warnings saying 'Error evaluating expressions.

Here is the code snippet:

```
var Vue = require('vue');
module.exports = new Vue({
    el: '#college-view-page',
    data: {
        college: {},
    },
    ready: function () {
        this.fetchCollege();
    },
    methods: {
        fetchCollege: function () {
            $.ajax({
                url: this.$$.collegeId.getAttribute('url'),
                method: 'GET',
                statusCode: {
                    200: function (response) {
                        this.college = response;
                    }.bind(this)
                }
            });
        }
    },
    filters: {
        fixUrl: require('../../filters/fixUrl')
    }
});
```

In my view file this line throws a warning:

```
<td>{{ college.status.name }}</td>
```

```
[Vue warn]: Error when evaluating expression "college.status.name"
```

What is the work around? Do I need to iterate over the response object and set properties of the object one by one using $set?

**Comment by yyx990803** [comment]
@abhijitghogre you can actually safely ignore this warning. You can suppress expression evaluation warnings by setting `Vue.config.warnExpressionErrors = false`.

Or, initialize your data more specifically (although verbose, this has the benefits that you know what shape the ajax response looks like):

``` js
data: {
  college: {
    status: {
      name: ''
    }
  }
}
```

**Comment by abhijitghogre** [comment]
Thanks Evan!

**Comment by abhijitghogre** [comment]
Evan,

I used this to set image src:

```html
<img v-attr="src: coverImagePath">
```

When I change 'coverImagePath' model the image doesn't update. Any workaround?

**Comment by mark-hahn** [comment]
That should work.  Can you show your code that creates and/or passes
the `coverImagePath`
value in?

**Comment by abhijitghogre** [comment]
coverImage path should get computed automatically when college changes right?

This is where 'college' model is changed.
```javascript
methods: {
          fetchCollege: function () {
                $.ajax({
                    url: $('.showCollegeRoute').val(),
                    method: 'GET',
                    statusCode: {
                        200: function (response) {
                            this.college = response;
                        }.bind(this)
                    }
                });
            }
}
```

And this is how it is declared in data:
```javascript

data: {
            college: {}
        },

```

**Comment by yyx990803** [comment]
@abhijitghogre Looks like it should work. Open an issue at vuejs/Discussion and make a fiddle reproduction please.

**Comment by abhijitghogre** [comment]
Here is the fiddle: https://jsfiddle.net/pc2gg27y/5/

I have also created issue on Discussion repo #216

**Comment by lusty15** [comment]
I'm new to Vue, and attempting to render a JSON object returned by PHP using it's JSON_ENCODE function.  Is there a simple way of loading this data into a list?

**Comment by azamat-sharapov** [comment]
@lusty15 has 38 participants and is closed, so that people don't spam with questions. I think it's best to ask questions like yours on vuejs/Discussion or in gitter channel of yyx990803/vue repo. 

As for you question, there is vuejs/vue-resource plugin for AJAX calls, you can use it to fetch JSON from server.

**Comment by lusty15** [comment]
Thank you.

**Comment by nicklaw5** [comment]
@yyx990803 Hi Evan.

For starters really loving Vue.js. Great job!

To my question: Is it possible to "trigger" jQuery-like events? I posted [this question on StackOverflow](http://stackoverflow.com/questions/31917889/is-it-possible-to-trigger-events-using-vue-js), then came across this thread and thought it might help others who come across the same situation.

**Comment by simplesmiler** [comment]
@nicklaw5 why do you need the event in your method?

If it carries some useful information (target element, or click coordinates), then I'd say it's better to create a dedication method with corresponding signature, and call this method from the click handler.

The most common use case for having the event is to `event.preventDefault()`, but you don't need that if you call your method from the code, because you don't have the event in the first place.

**Comment by sl-codenine** [comment]
Hi, according to discussion from [here](https://github.com/yyx990803/vue/issues/96#issuecomment-37870300) my [jsfiddle](http://jsfiddle.net/plunntic/g7dteuve/) should work, but it looks like those native events don't trigger data sync in vue?

**Comment by yyx990803** [comment]
@sl-codenine v-model listens to `input` event by default, and only listens to `change` in lazy mode: http://jsfiddle.net/yyx990803/g7dteuve/2/

**Comment by JosephSilber** [comment]
@yyx990803 I've run into this before.

Any reason the default non-lazy `v-model` doesn't *also* listen to `change`?

**Comment by sl-codenine** [comment]
@yyx990803 Cool, that should work for me, thanks :)

**Comment by thecotne** [comment]
is there any way to change several variables without rendering intermediate stages?
it makes UI very slow and some times it is just wrong

**Comment by yyx990803** [comment]
@thecotne Vue's default behavior buffers the rendering until next event loop, so what you described should never happen. If you are running into problems, create an issue with a reproduction.

**Comment by deefour** [comment]
I'm having trouble understanding how a computed property is observed from a child via `props` vs. `inherit`. Given the following parent view:

**Parent Vue**

```javascript
var Parent = Vue.extend({
  
  data() {
    return {
      orders: [].
      tab: 'current'
    };
  }

  computed: {

    scoped_orders() {
      return this[this.tab];
    },
    
    completed() {
      return this.orders;
    }
    
    future() {
      return this.orders;
    }
    
  },
  
  components: {
    child: Child
  }

});
```

why **does** the following work *(using `'inherit'`)*

**Child Vue**

```javascript

  var Child = Vue.extend({

    template: require('some/template/path.mustache'),

    inherit: true,
    
  });
```

but the following **does not** *(using `props`)*:

**Child Vue**

```javascript

  var Child = Vue.extend({

    template: require('orders/index.mustache'),

    props: [ 'scoped_orders' ]
    
  });
```

Specifically, I have a template like this for the child

```html
<div v-repeat="scoped_orders">
   <!-- ... -->
```
 
 and when using `props` to fetch the computed `scoped_orders` property, I get the following error
 
 ```
[Vue warn]: Error when evaluating expression "scoped_orders.length".
TypeError: Cannot read property 'length' of undefined
 ```

Using `inherit` instead of `props`, the child view observes changes to `scoped_orders` without a problem.

*Worth mentioning, `orders` on the parent is updated via an Ajax response with a set of data, but is initially empty. The error occurs before the Ajax has time to finish.*

**Comment by mark-hahn** [comment]
I've never use a prop with an underscore.  I'm not sure what it would
convert into for the hyphen-case version.  Could that be it?

**Comment by Voyga** [comment]
What is the exact difference between `vm.$data.prop` and `vm.prop`? Under what circumstances should we use `$data`?

**Comment by yyx990803** [comment]
@Voyga see http://vuejs.org/guide/#Model

`vm.prop` is simply a getter/setter proxy to `vm.$data.prop`.

**Comment by chrisvfritz** [comment]
Is `v-model` really supported in IE9? It looks like [the `input` event doesn't work in that browser](http://stackoverflow.com/questions/6382389/oninput-in-ie9-doesnt-fire-when-we-hit-backspace-del-do-cut#answer-13780757).

**Comment by chenjie-cnooc** [comment]
what's the dirty check using html string? when I use the Vue.Router,I can't change to other hashurl.

**Comment by sabrinaluo** [comment]
@bpierre 
hi, thanks for your v-attr hint, it works when the attribute is simple.
do you have any idea how to make this work? or do you have any other way to make it work? thank you :)

```
<g v-for="d in data" transform="translate({{$index}}, 0)">
```

**Comment by curtisblackwell** [comment]
I have an array of data in my Vue instance with Unix timestamps for each element. I'm looping over the data with `v-for` and trying to filter it with a `v-model` `input`.

I want people to be able to specify a range to filter by in plain English, so I'm trying to do this:

- [x] Send the plain English data to the server where PHP parses it and sends back some JSON.
- [x] Use that JSON to filter the array sent to the `Vue.filter()` method.
- [ ] Return the filtered array so that the view updates.

It seems as though it's not possible b/c if I return the array (or part of it) before the AJAX call, the filtering "works." However, if I return within the `this.$http.get()` request, it seems as though Vue stopped paying attention. I tried playing around with `debounce` a bit, but to no avail.

Should this work? Is there something I should do instead?

**Comment by posva** [comment]
@curtisblackwell 
Ask the question on the [forums](forum.vuejs.org) or the [gitter chat](https://gitter.im/vuejs/vue). This is no longer used

**Comment by curtisblackwell** [comment]
@posva will do. Why not close/lock this then?

**Comment by blake-newman** [comment]
Locked: 

Please ask questions or post informatino on the [forums](forum.vuejs.org) or the [gitter chat](https://gitter.im/vuejs/vue). This is issue is no longer in use.
